content/blog/2022/2022-09-02-junit5.md

```diff
@@ -0,0 +1,601 @@
+		 <groupId>org.junit.jupiter</groupId>  
+		 <artifactId>junit-jupiter-api</artifactId>  
+		 <version>5.9.0</version>  
+		 <scope>test</scope>  
+	 </dependency>
+ </dependencies>  
+  
+<build>  
+ <finalName>maven-unit-test</finalName>  
+	 <plugins> 
+		 <plugin> 
+			 <groupId>org.apache.maven.plugins</groupId>  
+			 <artifactId>maven-surefire-plugin</artifactId>  
+			 <version>3.0.0-M7</version>  
+		 </plugin> 
+	 </plugins>
+ </build>
+``` 
+
+### @Test
+Adding this annotation to a public void method allows us to simply run that method as a test.
+```java
+import org.junit.jupiter.api.Test;  
+import static org.junit.jupiter.api.Assertions.assertEquals;  
+  
+public class DogTest {  
+  @Test  
+  public void testBark() {  
+        String expectedString = "woof";  
+	    assertEquals(expectedString, "woof");  
+  }  
+}
+```
+Here we use the `assertEquals` method to assert that the two strings are equal, we will cover it in more detail later.
+Now, to run the tests all we need to do is run the maven command `mvn test`.
+
+We should see the test run and pass:
+```
+[INFO] Running DogTest
+[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 s - in DogTest
+[INFO] 
+[INFO] Results:
+[INFO] 
+[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
+```
+
+Let's check what happens when the test fails.
+```java
+public class DogTest {  
+	@Test  
+	public void barkFailure() {  
+	    String expectedString = "Meow";  
+	    assertEquals(expectedString, "Woof");  
+	}
+}
+```
+Then we will see
+```
+[INFO] Running DogTest
+[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.066 s <<< FAILURE! - in DogTest
+[ERROR] DogTest.barkFailure  Time elapsed: 0.036 s  <<< FAILURE!
+org.opentest4j.AssertionFailedError: expected: <Meow> but was:
+```
+
+### @BeforeAll
+The method annotated with it has to be a static method and it will run once before any test is run.
+
+```java
+public class DogTest {  
+  
+  @BeforeAll  
+  public static void init() {  
+        System.out.println("Doing stuff");  
+  }  
+  @Test  
+  public void testBark() {  
+        String expectedString = "woof";  
+		assertEquals(expectedString, "woof");  
+	    System.out.println("WOOF!");  
+  }  
+}
+```
+This will output
+```
+Doing stuff
+WOOF!
+```
+
+### @BeforeEach
+The name says it all! The method annotated with it will run once before every test method annotated with @Test.
+
+```java
+public class DogTest {  
+	@BeforeEach  
+	public void doEach() {  
+	    System.out.println("Hey Doggo");  
+	}  
+	  
+	@Test  
+	public void testBark1() {  
+	    String expectedString = "woof1";  
+		assertEquals(expectedString, "woof1");  
+	    System.out.println("WOOF => 1");  
+	}  
+	  
+	@Test  
+	public void testBark2() {  
+	    String expectedString = "woof2";  
+	    assertEquals(expectedString, "woof2");  
+	    System.out.println("WOOF => 2");  
+	}
+}
+```
+This will output
+```
+Hey Doggo
+WOOF => 1
+Hey Doggo
+WOOF => 2
```

Comment:
The log output implies that the order of tests is guaranteed, which is not the case. Maybe add a note box here explaining that test order is not guaranteed? See https://reflectoring.io/contribute/writing-guide/#asides